### ISSUE TITLE
DeepChem Common Errors

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#2513]
When using DeepChem, there are some classes of common errors I've observed that don't have great error messages. I'm starting this thread up for us to start documenting some patterns which hopefully can lead to improved error messages:

- **Using the wrong featurization:** When copy/pasting example code, you'll often do things like using ECFP featurizations with graph convs which will yield a very confusing TensorFlow error. This ideally should have a natural error message suggesting the user check their featurization.
- **Padding:** The generators in KerasModel pad by default. When using custom models, this often confuses newcomers since they enter in a dataset of say 100 elements with batch size 64 and are very surprised to receive back 128 predictions. Some warning should be emitted along with a suggested fix

I'll add on more errors here as I come across them. Please add on your own observed confusing DeepChem behavior!

**Comment by manangoel99** [comment]
Hi! I'd like to start contributing. Can I take this up?

**Comment by rbharath** [comment]
@manangoel99 You're welcome to take a crack at it! I'd suggest getting starting by documenting errors you run into as you start working through the DeepChem tutorials

**Comment by manangoel99** [comment]
I went through the tutorials. A couple of things that I think could help

- The error with using ECFP with GraphConv, a more straightforward error message is required which can be done by adding a validation step before actually trying to fit the model.
- There's a huge warning message when the ```GraphConvModel``` is fit. Can this be fixed somehow?
```
UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Reshape_14:0", shape=(384,), dtype=int32), values=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Reshape_13:0", shape=(384, 64), dtype=float32), dense_shape=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Cast_4:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
```
- Is there a way to print the metrics on the training data like ```ValidationCallback```
- There are some deprecation warnings in the tutorials themselves like using the ```split``` argument in Dataloader classes and pandas cannot read xlsx files without the ```openpyxl``` package. Would it be okay if I fix these?
- I couldn't recreate the padding error. I tried taking a custom model and used generators from ```dc.molnet.load_delaney```, trained the model on the training set and used all the possible predict functions but it gave the correct batch sizes without padding.

**Comment by manangoel99** [comment]
@rbharath What do you think?

> I went through the tutorials. A couple of things that I think could help
> 
> * The error with using ECFP with GraphConv, a more straightforward error message is required which can be done by adding a validation step before actually trying to fit the model.
> * There's a huge warning message when the `GraphConvModel` is fit. Can this be fixed somehow?
> 
> ```
> UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Reshape_14:0", shape=(384,), dtype=int32), values=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Reshape_13:0", shape=(384, 64), dtype=float32), dense_shape=Tensor("gradient_tape/private__graph_conv_keras_model_4/graph_pool_9/Cast_4:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
> ```
> 
> * Is there a way to print the metrics on the training data like `ValidationCallback`
> * There are some deprecation warnings in the tutorials themselves like using the `split` argument in Dataloader classes and pandas cannot read xlsx files without the `openpyxl` package. Would it be okay if I fix these?
> * I couldn't recreate the padding error. I tried taking a custom model and used generators from `dc.molnet.load_delaney`, trained the model on the training set and used all the possible predict functions but it gave the correct batch sizes without padding.

**Comment by rbharath** [comment]
@manangoel99 Your suggestions sound great! Do you want to join one of our developer calls to discuss how to potentially help fix some of these issues?

**Comment by manangoel99** [comment]
Sure! I'll join tomorrow's meeting.

**Comment by manangoel99** [comment]
> * There's a huge warning message when the `GraphConvModel` is fit. Can this be fixed somehow?

I traced this warning down to the `Reshape` calls in `_GraphConvKerasModel` in `deepchem/models/graph_models.py`. I found the following link which explains why this could be happening
 https://stackoverflow.com/questions/35892412/tensorflow-dense-gradient-explanation 
I couldn't really find a replacement for the keras `Reshape` layer so I'm not quite sure how to fix this.

**Comment by manangoel99** [comment]
> * Is there a way to print the metrics on the training data like `ValidationCallback`

Apologies for listing this. I found that there is already an option for this when I took a much closer look.

**Comment by peastman** [comment]
We'll hopefully be converting GraphConvModel to PyTorch soon.  That will eliminate those messages.

**Comment by manangoel99** [comment]
> We'll hopefully be converting GraphConvModel to PyTorch soon. That will eliminate those messages.

That's awesome! I'd love to help out
